src/v/iceberg/compatibility.h

```diff
@@ -40,4 +40,33 @@ namespace iceberg {
 type_check_result
 check_types(const iceberg::field_type& src, const iceberg::field_type& dest);
 
+/**
+ * annotate_schema_transform - Answers whether a one schema can "evolve" into
+ * another, structurally.
+ *
+ * Traverse the schemas in lockstep, marking nested fields in one of the
+ * following ways:
+ *   - for a field from the source schema:
+ *     - the field does not exist in the dest schema (i.e. drop column)
+ *     - a backwards compatible field exists in the dest schema (update column)
+ *   - for a field from the dest schema:
+ *     - the field receives its ID from a compatible field in the source schema
+ *     - the field does not appear in the source schema (i.e. add column)
+ *
+ * Considerations:
+ *   - This function modifies nested_field metadata, but ID updates are deferred
+ *     to validate_schema_transform (see below). This is mainly to allow
+ *     constness throughout the traversal, ensuring that we do not make
+ *     inadvertent structural changes to either input struct.
+ *
+ * @param source - the schema we want to evolve _from_ (probably the current
+ *                 schema for some table)
+ * @param dest   - the proposed schema (probably extracted from some incoming
+ *                 record)
+ *
+ * @return std::nullopt (indicating success), or an error code
+ */
+schema_transform_result
+annotate_schema_transform(const struct_type& source, const struct_type& dest);
```

Comment:
Just a thought, this seems like a thing we could fuzz test

Reply:
yeah i agree. probably should. noted for a follow up 🙂